app/src/main/java/com/github/polypoly/app/network/RemoteDB.kt

```diff
@@ -53,123 +47,191 @@ class RemoteDB(
         return valuePromise
     }
 
-    override fun <T : Any> getValue(key: String, clazz: KClass<T>): CompletableFuture<T> {
-        return getValueAndThen(key, { data, valuePromise ->
-            valuePromise.complete(data.getValue(clazz.java)!!)
-        }, { valuePromise ->
-            valuePromise.completeExceptionally(NoSuchElementException("No value found for key <$key>"))
-        })
-    }
+    /**
+     * First gets the [DataSnapshot] associated to the given [absoluteKey] and then composes it
+     * with the corresponding given future (onSuccess if the data exists, onFailure else)
+     * @return a future composed with [onSuccess] or [onFailure]
+     */
+    private fun <T> getDataAndThen(
+        absoluteKey: String,
+        onSuccess: (DataSnapshot) -> CompletableFuture<T>,
+        onFailure: CompletableFuture<T>
+    ): CompletableFuture<T> {
+        val waitingFuture = CompletableFuture<CompletableFuture<T>>()
 
-    override fun <T : Any> getAllValues(key: String, clazz: KClass<T>): CompletableFuture<List<T>> {
-        return getValueAndThen(key, { data, valuesPromise ->
-            val users = ArrayList<T>()
-            for (child in data.children) {
-                users.add(child.getValue(clazz.java)!!)
+        rootRef.child(absoluteKey).get().addOnSuccessListener { data ->
+            if(!data.exists()) {
+                waitingFuture.complete(onFailure)
+            } else {
+                waitingFuture.complete(onSuccess(data))
             }
-            valuesPromise.complete(users)
-        }, { promise -> promise.complete(listOf()) })
+        }.addOnFailureListener(waitingFuture::completeExceptionally)
+
+        return waitingFuture.thenCompose { future -> future }
+    }
+
+    /**
+     * @return a future with true iff the given [absoluteKey] is in the DB
+     */
+    private fun absoluteKeyExists(absoluteKey: String): CompletableFuture<Boolean> {
+        return getDataAndThen(
+            absoluteKey,
+            { CompletableFuture.completedFuture(true) },
+            CompletableFuture.completedFuture(false)
+        )
     }
 
-    override fun getAllKeys(parentKey: String): CompletableFuture<List<String>> {
-        return getValueAndThen(
-            parentKey,
-            { data, keysPromise ->
+    /**
+     * Gets the reference of a given key and if it exists, executes the given action with that reference
+     * @return a future with true if the action was successfully executed, false else or if no value was found
+     */
+    private fun getRefAndThen(absoluteKey: String, action: (DatabaseReference) -> Unit): CompletableFuture<Boolean> {
+        return getDataAndThen(
+            absoluteKey,
+            { data ->
+                action(data.ref)
+                CompletableFuture.completedFuture(true)
+            },
+            CompletableFuture.completedFuture(false)
+        )
```

Comment:
Have seen this above and below, but you can add the import:
import java.util.concurrent.CompletableFuture.completedFuture
in order to avid repeating CompletableFuture (Intellij will help you with the yellow marks on the sidebar)